modules/core/src/test/java/org/apache/ignite/internal/processors/cache/SafeLogTxFinishErrorTest.java

```diff
@@ -0,0 +1,163 @@
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setCacheConfiguration(
+            new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+                .setAffinity(new RendezvousAffinityFunction(false, 32))
+                .setAtomicityMode(TRANSACTIONAL)
+        );
+
+        if (removeFailureHandler)
+            cfg.setFailureHandler(null);
+
+        return cfg;
+    }
+
+    /**
+     * Checking that {@link IgniteTxAdapter#logTxFinishErrorSafe} is performed
+     * without errors with FailureHandler.
+     *
+     * @throws Exception If any error occurs.
+     */
+    @Test
+    public void testSafeLogTxFinishErrorWithFailureHandler() throws Exception {
+        checkSafeLogTxFinishError();
+    }
+
+    /**
+     * Checking that {@link IgniteTxAdapter#logTxFinishErrorSafe} is performed
+     * without errors without FailureHandler.
+     *
+     * @throws Exception If any error occurs.
+     */
+    @Test
+    public void testSafeLogTxFinishErrorWithoutFailureHandler()throws Exception {
+        removeFailureHandler = true;
+
+        checkSafeLogTxFinishError();
+    }
+
+    /**
+     * Checking that {@link IgniteTxAdapter#logTxFinishErrorSafe} is performed
+     * without errors.
+     *
+     * @throws Exception If any error occurs.
+     */
+    private void checkSafeLogTxFinishError() throws Exception {
+        IgniteEx igniteEx = startGrid(0);
+
+        try (Transaction transaction = igniteEx.transactions().txStart()) {
+            IgniteCache<Object, Object> cache = igniteEx.cache(DEFAULT_CACHE_NAME);
+
+            cache.put(1, 1);
+
+            Collection<IgniteInternalTx> activeTxs = igniteEx.context().cache().context().tm().activeTransactions();
+
+            assertEquals(1, activeTxs.size());
+
+            GridNearTxLocal activeTx = (GridNearTxLocal)activeTxs.iterator().next();
+
+            AtomicBoolean containsFailedCompletingTxInLog = new AtomicBoolean();
+
+            /*
+             * When GridNearTxLocal#toString() is called, return string
+             * contains substring "duration", which is calculated as
+             * current time minus start time of the transaction.
+             * For test, we need so that when call GridNearTxLocal#toString(),
+             * we get the same string with the same substring "duration".
+             */
+            AtomicLong curTimeMillis = new AtomicLong();
+
+            log.registerListener(logStr -> {
+                /*
+                 * Restore the current time to correctly call 
+                 * GridNearTxLocal#toString().
+                 */
+                setFieldValue(IgniteUtils.class,"curTimeMillis", curTimeMillis.get());
```

Comment:
Tests are flaky because of gap between setFieldValue() and activeTx.toString(). In order to verify that it's possible to add Thread.sleep(10) after setFieldValue().